File: day-04a/src/main.py
from collections.abc import Iterator
from dataclasses import dataclass


@dataclass(frozen=True, slots=True, kw_only=True)
class Card:
    card_id: int
    winning_numbers: frozenset[int]
    numbers_we_have: frozenset[int]

    def total_points(self) -> int:
        intersection = self.winning_numbers & self.numbers_we_have
        if not intersection:
            return 0
        return 2 ** (len(intersection) - 1)


def parse_input(filename: str) -> Iterator[Card]:
    with open(filename) as f:
        for card_id, line in enumerate(f, start=1):
            _, _, data = line.partition(": ")
            left, _, right = data.partition(" | ")
            winning_numbers = frozenset(map(int, left.split()))
            numbers_we_have = frozenset(map(int, right.split()))
            yield Card(
                card_id=card_id,
                winning_numbers=winning_numbers,
                numbers_we_have=numbers_we_have
            )


def solve(filename: str) -> int:
    return sum(card.total_points() for card in parse_input(filename))


def main() -> None:
    print(solve("input.txt"))


if __name__ == "__main__":
    main()

File: day-04a/src/main.rs
use std::collections::HashSet;
use std::fs::read_to_string;

struct Card {
    winning_numbers: HashSet<u32>,
    numbers_we_have: HashSet<u32>,
}

impl Card {
    fn total_points(&self) -> u32 {
        let intersection = self.winning_numbers.intersection(&self.numbers_we_have);
        match intersection.count() {
            0 => 0,
            number => 2_u32.pow((number as u32) - 1),
        }
    }
}

fn parse_input(filename: &str) -> Vec<Card> {
    let mut cards = vec![];
    for line in read_to_string(filename).unwrap().lines() {
        let [_, data] = line.split(": ").collect::<Vec<&str>>()[..] else {
            panic!()
        };
        let [left, right] = data.split(" | ").collect::<Vec<&str>>()[..] else {
            panic!()
        };
        let winning_numbers =
            HashSet::<u32>::from_iter(left.split_whitespace().map(|n| n.parse::<u32>().unwrap()));
        let numbers_we_have =
            HashSet::<u32>::from_iter(right.split_whitespace().map(|n| n.parse::<u32>().unwrap()));
        cards.push(Card {
            winning_numbers,
            numbers_we_have,
        })
    }
    cards
}

fn solve(filename: &str) -> u32 {
    parse_input(filename).iter().map(|c| c.total_points()).sum()
}

fn main() {
    println!("{}", solve("input.txt"));
}
